canopen/profiles/p402.py

```diff
@@ -0,0 +1,144 @@
+    0x07: 'SWITCHED ON',
+    0x0F: 'OPERATION ENABLED',
+    0x02: 'QUICK STOP ACTIVE'
+}
+
+class Node402(Node):
+    """A CANopen CiA 402 profile slave node.
+
+    :param int node_id:
+        Node ID (set to None or 0 if specified by object dictionary)
+    :param object_dictionary:
+        Object dictionary as either a path to a file, an ``ObjectDictionary``
+        or a file like object.
+    :type object_dictionary: :class:`str`, :class:`canopen.ObjectDictionary`
+    """
+
+    def __init__(self, node_id, object_dictionary):
+        super(Node402, self).__init__(node_id, object_dictionary)
+        self.powerstate_402 = PowerStateEngine(self)
+        self.powerstate_402.network = self.network
+
+    def setup_402_state_engine(self):
+        # setup TPDO1 for this node
+        # TPDO1 will transmit the statusword of the 402 power state engine
+        # first read the current PDO setup and only change TPDO1
+        print(self.nmt.state)
+        self.nmt.state = 'PRE-OPERATIONAL'
+        self.pdo.read()
+        self.pdo.tx[1].clear()
+        # Use register as to stay manufacturer agnostic
+        self.pdo.tx[1].add_variable(0x6041)
+        # add callback to listen to TPDO1 and change 402 state
+        self.pdo.tx[1].add_callback(self.powerstate_402.on_PDO1_callback)
+        self.pdo.trans_type = 255
+        self.pdo.tx[1].enabled = True
+        self.pdo.save()
+        self.nmt.state = 'OPERATIONAL'
+
+class PowerStateEngine(object):
+    """A CANopen CiA 402 Power State engine. Listens to state changes
+    of the DS402 Power State engine by means of TPDO 1 Statusword.
+
+    - Controlword 0x6040 causes transitions
+    - Statusword 0x6041 gives the current state
+
+    """
+
+    def __init__(self, node):
+        self.id = node.id
+        self.node = node
+        self._state = 'NOT READY TO SWITCH ON'
+
+    @staticmethod
+    def on_PDO1_callback(mapobject):
+        # this function receives a map object.
+        # this map object is then used for changing the
+        # Node402.PowerstateEngine._state
+        #
+        # I would have preferred to retrieve the statusword by means of the
+        # register as to make this ParameterName string (vendor) agnostic like:
+        # statusword = mapobject.pdo_node.tx[1][0x6041].raw
+        #
+        # Since I haven't found a way to do this the ParameterName is first
+        # retrieved from the OD, and used for extracting the data in the Map.
+        description = mapobject.pdo_node.node.object_dictionary[0x6041].name
+        statusword = mapobject.pdo_node.tx[1][description].raw
+        for key, value in POWER_STATES_402.iteritems():
+    		# check if the value after applying the bitmask (value[0])
+    		# corresponds with the corresponding value to determine
+    		# the current status
+            bitmaskvalue = statusword & value[0]
+            if bitmaskvalue == value[1]:
+                mapobject.pdo_node.node.powerstate_402._state = key
```

Comment:
I struggled a lot with adding this function to the callback. It looked like calling self.something blocks the code somehow while a static function worked. So the workaround was to access _state from the passed object. I there is a better way i have no problem changing to that.